Assets/Virtual Agents Framework/Runtime/Scripts/ScheduleBasedExecution/TaskBundle.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using i5.VirtualAgents.AgentTasks;
+using UnityEngine;
+
+namespace i5.VirtualAgents
+{
+    /// <summary>
+    /// A task which consists of multiple subtasks.
+    /// It allows for checking of preconditions and then executing a sequence of tasks.
+    /// </summary>
+    public class TaskBundle : AgentBaseTask
+    {
+        /// <summary>
+        /// Creates an empty TaskBundle
+        /// </summary>
+        public TaskBundle()
+        {
+            State = TaskState.Waiting;
+            TaskQueue = new List<AgentBaseTask>();
+            Preconditions = new List<Func<bool>>();
+        }
+
+        /// <summary>
+        /// Creates a TaskBundle with a list of subtasks
+        /// </summary>
+        /// <param name="tasks"></param>
+        public TaskBundle(List<AgentBaseTask> tasks)
+        {
+            State = TaskState.Waiting;
+            TaskQueue = tasks;
+            Preconditions = new List<Func<bool>>();
+        }
+
+        /// <summary>
+        /// Creates a TaskBundle with a list of tasks and a list of preconditions
+        /// </summary>
+        /// <param name="tasks"></param>
+        /// <param name="preconditions"></param>
+        public TaskBundle(List<AgentBaseTask> tasks, List<Func <bool>> preconditions)
+        {
+            State = TaskState.Waiting;
+            TaskQueue = tasks;
+            Preconditions = preconditions;
+        }
+
+        /// <summary>
+        /// List of tasks to be part of the bundle
+        /// </summary>
+        private List<AgentBaseTask> TaskQueue { get; set; }
+
+        /// <summary>
+        /// List of conditions to be met before execution of tasks
+        /// </summary>
+        private List<Func<bool>> Preconditions { get; set; }
+
+        /// <summary>
+        /// Check for preconditions and start the execution of all subtasks in sequence
+        /// </summary>
+        /// <param name="executingAgent"></param>
+        public override void StartExecution(Agent executingAgent)
+        {
+            State = TaskState.Running;
+            if (CheckPreconditions())
+            {
+                executingAgent.StartCoroutine(ExecuteTasks(executingAgent));
+            }
+        }
+
+        /// <summary>
+        /// Execute all tasks in the task queue. If a task fails, the whole bundle fails. Note, that checking of preconditions is not part of this method.
+        /// </summary>
+        /// <param name="executingAgent"></param>
+        private IEnumerator ExecuteTasks(Agent executingAgent)
```

Comment:
Would it be possible to use/overwrite the Tick() method of the base task instead of executing the task as a co-routine? The tick of an active task gets called every frame, so we could refactor the parts. This would also give us the flexibility in the future to, e.g., suspend tasks by stopping calling the tick method.

You could also re-use code from the AgentTaskQueue and maybe even from the AgentTaskManager for this, as these classes implement a very similar TaskQueue logic. Maybe we can create the solution in a way that a task bundle contains its own AgentTaskManager that runs the contained tasks. This way, the logic for selecting the next task and deciding when to switch to the next task is consistent with the overall task queue of the agent as its handled by the same class.

Reply:
The TaskBundles now use a TaskManager to schedule the tasks. And we override EvaluateTaskState() instead of using a co-routine. See [dea7a84](https://github.com/rwth-acis/Virtual-Agents-Framework/pull/17/commits/dea7a84c2848818a57c7d4a0b484fd5d0b219200).